src/QsCompiler/SyntaxProcessor/CapabilityInference.fs

```diff
@@ -0,0 +1,240 @@
+﻿module Microsoft.Quantum.QsCompiler.SyntaxProcessing.CapabilityInference
+
+open Microsoft.Quantum.QsCompiler
+open Microsoft.Quantum.QsCompiler.DataTypes
+open Microsoft.Quantum.QsCompiler.Diagnostics
+open Microsoft.Quantum.QsCompiler.ReservedKeywords.AssemblyConstants
+open Microsoft.Quantum.QsCompiler.SyntaxTokens
+open Microsoft.Quantum.QsCompiler.SyntaxTree
+open Microsoft.Quantum.QsCompiler.Transformations
+open Microsoft.Quantum.QsCompiler.Transformations.Core
+open Microsoft.Quantum.QsCompiler.Transformations.SearchAndReplace
+
+/// A syntactic pattern that requires a runtime capability.
+type private Pattern =
+    /// A return statement in the block of an if statement whose condition depends on a result.
+    | ReturnInResultConditionedBlock of Range QsNullable
+
+    /// A set statement in the block of an if statement whose condition depends on a result, but which is reassigning a
+    /// variable declared outside the block. Includes the name of the variable.
+    | SetInResultConditionedBlock of string * Range QsNullable
+
+    /// An equality expression inside the condition of an if statement, where the operands are results
+    | ResultEqualityInCondition of Range QsNullable
+
+    /// An equality expression outside the condition of an if statement, where the operands are results.
+    | ResultEqualityNotInCondition of Range QsNullable
+
+/// The level of a runtime capability. Higher levels require more capabilities.
+type private Level = Level of int
+
+/// Returns the required runtime capability of the pattern, given whether it occurs in an operation.
+let private toCapability inOperation = function
+    | ResultEqualityInCondition _ -> if inOperation then RuntimeCapabilities.QPRGen1 else RuntimeCapabilities.Unknown
+    | ResultEqualityNotInCondition _ -> RuntimeCapabilities.Unknown
+    | ReturnInResultConditionedBlock _ -> RuntimeCapabilities.Unknown
+    | SetInResultConditionedBlock _ -> RuntimeCapabilities.Unknown
+
+/// The runtime capability with the lowest level.
+let private baseCapability = RuntimeCapabilities.QPRGen0
```

Comment:
It might be worth it to add a comment with the characteristics of each runtime capability in `ReservedKeywords.fs` where the names of the capabilities are defined. That way, when they are referenced, people can look at the definition and have a bit more context.